### DISCUSSION TITLE
How big does a server have to be to run Jamulus

### DISCUSSION THREAD

**Discussion by duttonphilip** [jamulussoftware/jamulus#1133] [Documentation]
The question...
Is our server powerful enough for Jamulus? We have 1 intel Xeon Vcore CPU and 1024 MB of RAM and a bandwidth of 100 GB.

The background....
We have rented a cloud based server in our city, installed Jamulus and have managed some singing rehearsals with about 25 people. The latency is excellent for some and a bit difficult for others but understand that this all to do with the clients' set-ups. What we have noticed though is that when everyone is singing the performance gets worse and wondered if that could be because the server cannot cope with this number of inputs and outputs. Should we upgrade it?

**Comment by 38github** [comment]
Oh, yes it will be enough. I ran a Jamulus server on a laptop using Ubuntu 20.04 server installed on a USB where the fifteen year old Celeron cpu only get 300 point in passmark benchmarks (really slow). It handled at least ten clients without any issue at all and the RAM usage was between 100-150 MiB in idle and I can't remember that it ever used more than 300 MiB.

My experience is that using Jamulus on "bare metal" yields in many cases the best performance. Now I run my Jamulus servers in FreeBSD jails and it is super stable. Probably the most stable servers I have ever connected to. Before that I tried Linux in FreeBSD bhyve virtual machnes but the ping was not really stable which affected the jitter buffer and audio artifacts. An internet friend of mine tried using the docker Jamulus in a virtual machine in the cloud running Debian and the ping was fluctuating like crazy. He then stopped using the docker version and installed a native Debian build and the ping improved a lot but was still fluctuating some. The same I experience from many servers that in their welcomne message say that they use a cloud server.

**Comment by ann0see** [comment]
Hi! Thanks for this question. I think we should have a bit more discussion about this because I‘d say this must be documented somewhere on the website. I‘ll therefore move it to a discussion on the website. 

https://github.com/jamulussoftware/jamuluswebsite/issues/54 Is related

**Comment by gilgongo** [comment]
Just been replying to this discussion - https://github.com/jamulussoftware/jamulus/discussions/1235

If everyone ran bare metal servers, then perhaps a useful "sizing" document could be produced, but as you describe, there are so many different configurations on the same base specification (CPU, RAM) that you end up comparing apples and pears. Does a Docker image on a cloud host with notionally the same resources as a laptop on a home DSL connection perform better or worse? What about double-NATed ISPs? Real time kernels, etc etc? I think guestimation and experimentation are all you can really do. 

And as you also point out, most Jamulus servers aren't breaking a sweat.

**Reply by gene96817** [reply]
The best we can do is to use a reference configuration. That would establish the reference. Adding (turning on) features or complications will reduce capacity. At least we would all have a common reference point. It would be awesome to then be able to say: * turn on recording.   -X%, * use containers.       -Y%, or * user >K ms away.   -Z%.

Models are always (an over) simplifications. Properly done, models are really helpful.

**Reply by gilgongo** [reply]
But that's my point - you'd need to have reference configurations for every type of host, no? So bare metal, KVM, CT, Docker, and maybe others. And even if we did, I might have the exact same CT config as you, but my ISP is bursting my CPU, or RAM or bandwidth, and your's isn't. It's all just too chaotic to compare, isn't it?

**Comment by gene96817** [comment]
I would use bare metal with a Ubuntu and Jamulus  for areference configuration. All the other items add load to the bare metal. Every additional technology offers guidance on how efficient (i.e how little it consumes resources). Having a clean reference model is an essential foundation. If we have a reference model, it would be possible to understand which cloud strategies (from cloud vendors) are better for the server performance.

For me, I would expect my ISP to explain why they need to impose (or is it inflict)  the limitations, etc. Of course, some vendors have no interest in explaining why they make choices that affect their customers negatively.  Mostly the ISPs don't expect to have informed customers. I have another (not Jamulus) project, helping our local regulators understand how our ISPs are not delivering what the subscribers assume they are getting. (Yeah, the subscribers are getting only a fraction of what they are paying for.)
